README.md
[![Build Status](https://travis-ci.org/mk2366/WS2801python_rpi.svg?branch=master)](https://travis-ci.org/mk2366/WS2801python_rpi)
[![codecov](https://codecov.io/gh/mk2366/WS2801python_rpi/branch/master/graph/badge.svg)](https://codecov.io/gh/mk2366/WS2801python_rpi)
[![Version](https://img.shields.io/pypi/v/WS2801-RPI.svg?style=flat-square)](https://pypi.python.org/pypi/WS2801-RPI)
[![Implementation](https://img.shields.io/pypi/implementation/WS2801-RPI.svg?style=flat-square)](https://pypi.python.org/pypi/WS2801-RPI)
[![Status](https://img.shields.io/pypi/status/WS2801-RPI.svg?style=flat-square)](https://pypi.python.org/pypi/WS2801-RPI)
[![Pyversions](https://img.shields.io/pypi/pyversions/WS2801-RPI.svg?style=flat-square)](https://pypi.python.org/pypi/WS2801-RPI)
[![Format](https://img.shields.io/pypi/format/WS2801-RPI.svg?style=flat-square)](https://pypi.python.org/pypi/WS2801-RPI)
[![License](https://img.shields.io/pypi/l/WS2801-RPI.svg?style=flat-square)](https://pypi.python.org/pypi/WS2801-RPI)
[![Issues](http://img.shields.io/github/issues/mk2366/WS2801python_rpi.svg)]( https://github.com/mk2366/WS2801python_rpi/issues )


# WS2801python_rpi
## A python module connecting RPI with WS2801

It is based on the spidev module you can find [here](https://github.com/doceme/py-spidev).

You can use it like this:

```python
import WS2801_RPI as w
s=w.set_led_colors_buffer_list_multi_call

w.set_number_of_leds(100)                      # adjust to the number of leds in your project
w.hello()
w.s(1)                                         # to let led number 1 shine in white
w.flush()                                      # to communicate your settings to the led strip via spi
```

You can call set_leds as often as you like before flushing it out. And of course you can flush as often
as you need to create effects.

## Other convenience functions are:

* `set_number_of_leds(leds=128)`: Call this to set the number of LEDs on your strip for WS2801_RPI.
* `hello()`: I had many issues before I started to do a flush() with all LEDs white. This is a convenience function to let all LEDs shine 3 times for one second.
If you call it and nothing happens: There must be an issue either in program (Check `set_number_of_leds` and `set_spidev_bus_device`) or in your wiring
* `clear()`: turn off all leds. Remember you need to call `flush()` to see the effect.
* `get_led_colors_buffer_dict()`: Get all rgb values currently in the buffer as dictionary. This might differ from what you see on your LED strip.
* `set_led_colors_buffer_dict_multi_call()`: After you adjusted the dict of the function above: Write it back to the buffer. Don't forget `flush()`.
* `get_led_colors_buffer_list()`: Get back two lists. First is a list with integers for all of your LEDs (boring but needed for the set function) and a list of lists each with the three
values for rgb
* `set_led_colors_buffer_list_multi_call(pixels, rgb_values=[255, 255, 255])`: Set the colors for a list of LEDs (pixels). If you give less lists for RGB values than numbers for LEDs the last RGB value is assumed for the remaining LEDs. If you give only a list for rgb_values (like in the default) all LEDs in pixels are set to these values. If pixel is an int you are adressing one LED with this int. I know that this kind of multi is not for production but with this setup my daughter was happy and could play arround before doing all the python list topics.
* `set_max_speed_hz(hz)`: Default is 1MHz. You may change this but not below 1,5KHz.
* `flush()`: Send the data out to the LED strip via SPI
* `set_gamma()`: Set the gamma value. Formula is: `int(((rgb/255)**gamma)*255)`. Default is 2.1 which reduces the number of rgb values to approx. 187.
Set it to `1` you will have 255 values per color but you won't be able to see them. Your eye will turn many into white.
* `set_spidev_bus_device(bus=0, device=0)`: Change the default bus and device for WS2801_RPI
* `get_number_of_leds()`: You should know anyhow. But if you want to check how many LEDs WS2801_RPI is assuming call this function.

## Installation
`pip install WS2801_RPI`.

## Hint
You can raise the log level if you like to suppress warnings:
```python
import logging
logging.getLogger().setLevel(31)
```

## Testing
You can call `python setup.py test` to run the tests (if you clone the git repo onto your machine). I will update as soon as I have more understanding on the test infrastructure.

## Acknowledgement

* [Atom](https://atom.io/) was my editor of choice
* Formatting was done with [linter-pylama](https://atom.io/packages/linter-pylama)

## General remarks on WS2801:
* In the Data [Sheet](https://cdn-shop.adafruit.com/datasheets/WS2801.pdf) of WS2801 I found the following passage I didn't fully understand: `When the WS2801 receives total 24
clock rising edges, the WS2801 enters relay mode ...`. In addition one can read [this](https://electronics.stackexchange.com/a/307117) explanation. I found it helpful.
* WS2801 is using the raising flag of the clock to retrieve the data and expects clock being low for 500 microseconds between data transmissions. Hence there is no need to change
the mode of spidev.

## Would be glad if this code was used often in your projects. Feedback and pull requests are more than welcome.


CHANGES.txt
1.0.0, 22.12.2017  -- initial release
1.0.1, 2.1.2018    -- Version now attached to git tags
1.0.2, 2.1.2018    -- Samples


test/WS2801_RPI_test.py
"""Test module for Unit Tests of WS2801_RPI."""

import unittest
from unittest.mock import patch
import WS2801_RPI


@patch("WS2801_RPI.__spi")
class DataSetGetTestCase(unittest.TestCase):
    """Test the set/get methods in module WS2801_RPI."""

    def test_set_led_colors_buffer_list_multi_call(self, mock_spi):
        """Test set_led_colors_buffer_list_multi_call."""
        WS2801_RPI.set_led_colors_buffer_list_multi_call(1)
        WS2801_RPI.flush()
        WS2801_RPI.set_gamma(1)
        expected_result = ([255, 255, 255] + [0, 0, 0] *
                           (WS2801_RPI.get_number_of_leds() - 1))
        mock_spi.writebytes.assert_called_with(expected_result)
        with self.assertRaises(TypeError):
            WS2801_RPI.set_number_of_leds()
            WS2801_RPI.set_led_colors_buffer_list_multi_call("1")
        with self.assertRaises(TypeError):
            WS2801_RPI.set_number_of_leds()
            WS2801_RPI.set_led_colors_buffer_list_multi_call(1, 3)
        with self.assertRaises(TypeError):
            WS2801_RPI.set_number_of_leds()
            WS2801_RPI.set_led_colors_buffer_list_multi_call(["String"])
        with self.assertRaises(ValueError):
            WS2801_RPI.set_number_of_leds()
            WS2801_RPI.set_led_colors_buffer_list_multi_call([700])
        with self.assertRaises(ValueError):
            WS2801_RPI.set_number_of_leds()
            WS2801_RPI.set_led_colors_buffer_list_multi_call(700)
        with self.assertRaises(TypeError):
            WS2801_RPI.set_number_of_leds()
            WS2801_RPI.set_led_colors_buffer_list_multi_call([1, 3, 5],
                                                             [[255, 255, 255],
                                                             3])
        with self.assertRaises(TypeError):
            WS2801_RPI.set_number_of_leds()
            WS2801_RPI.set_led_colors_buffer_list_multi_call([1, 3, 5],
                                                             [[255, 255, 255],
                                                             [3]])
        with self.assertRaises(TypeError):
            WS2801_RPI.set_number_of_leds()
            WS2801_RPI.set_led_colors_buffer_list_multi_call([1, 3, 5],
                                                             [[255, 255, 255],
                                                             [3, 4, 6.7]])
        with self.assertRaises(ValueError):
            WS2801_RPI.set_number_of_leds()
            WS2801_RPI.set_led_colors_buffer_list_multi_call([1, 3, 5],
                                                             [[255, 255, 255],
                                                             [3, 4, 6700]])
        with self.assertRaises(TypeError):
            WS2801_RPI.set_number_of_leds()
            WS2801_RPI.set_led_colors_buffer_list_multi_call([1, 3, 5],
                                                             [255, 255]
                                                             )
        with self.assertRaises(TypeError):
            WS2801_RPI.set_number_of_leds()
            WS2801_RPI.set_led_colors_buffer_list_multi_call([1, 3, 5],
                                                             [255, 255, "String"]
                                                             )
        with self.assertRaises(ValueError):
            WS2801_RPI.set_number_of_leds()
            WS2801_RPI.set_led_colors_buffer_list_multi_call([1, 3, 5],
                                                             [255, 255, -3]
                                                             )

    def test_set_get_led_colors_buffer_dict(self, mock_spi):
        WS2801_RPI.set_gamma(1)
        WS2801_RPI.set_number_of_leds(1)
        d = WS2801_RPI.get_led_colors_buffer_dict()
        self.assertEqual(d, {1: {"red": 0, "green": 0, "blue": 0}})
        d[1]["red"] = 27
        WS2801_RPI.set_led_colors_buffer_dict_multi_call(d)
        WS2801_RPI.flush()
        mock_spi.writebytes.assert_called_with([27, 0, 0])
        with self.assertRaises(TypeError):
            WS2801_RPI.set_led_colors_buffer_dict_multi_call([1, "red"])
        d[-1] = {"green": 255}
        with self.assertRaises(ValueError):
            WS2801_RPI.set_led_colors_buffer_dict_multi_call(d)
        del(d[-1])
        WS2801_RPI.set_number_of_leds(2)
        d[2] = "No dictionary"
        with self.assertRaises(TypeError):
            WS2801_RPI.set_led_colors_buffer_dict_multi_call(d)
        d[2] = {"red": 256}
        with self.assertRaises(ValueError):
            WS2801_RPI.set_led_colors_buffer_dict_multi_call(d)
        d[2] = {"green": 256}
        with self.assertRaises(ValueError):
            WS2801_RPI.set_led_colors_buffer_dict_multi_call(d)
        d[2] = {"blue": 256}
        with self.assertRaises(ValueError):
            WS2801_RPI.set_led_colors_buffer_dict_multi_call(d)

    def test_set_spidev_bus_device(self, mock_spi):
        with self.assertRaises(TypeError):
            WS2801_RPI.set_spidev_bus_device("String", 0)
        WS2801_RPI.set_spidev_bus_device(0, 1)
        WS2801_RPI.flush()
        mock_spi.open.assert_called_with(bus=0, device=1)

    @patch("WS2801_RPI.time")
    def test_hello(self, mock_time, mock_spi):
        WS2801_RPI.hello()
        self.assertEqual(mock_spi.open.call_count, 6)
        self.assertEqual(mock_spi.writebytes.call_count, 6)

    @patch("WS2801_RPI.traceback")
    def test_flush(self, mock_traceback, mock_spi):
        mock_spi.open.side_effect = Exception("mock_spi raised an Excption")
        with self.assertRaises(Exception):
            WS2801_RPI.flush()

    def test_get_led_colors_buffer_list(self, mock_spi):
        WS2801_RPI.set_number_of_leds(5)
        WS2801_RPI.set_gamma(1)
        li = WS2801_RPI.get_led_colors_buffer_list()
        self.assertEqual(li, [[0, 0, 0]] * 5)


if __name__ == '__main__':
    unittest.main()


test/test_WS2801_RPI.py
"""Test module for Unit Tests of WS2801_RPI."""

import unittest
import sys
import WS2801_RPI
if sys.version_info >= (3,):
    from unittest.mock import patch
else:
    from mock import patch


@patch("WS2801_RPI.__spi")
class DataSetGetTestCase(unittest.TestCase):
    """Test the set/get methods in module WS2801_RPI."""

    def test_set_max_speed_hz(self, mock_spi):
        """Check whether one can set the speed"""
        WS2801_RPI.set_number_of_leds(5)
        with self.assertRaises(ValueError):
            WS2801_RPI.set_max_speed_hz(1000)
        WS2801_RPI.set_number_of_leds(5)
        WS2801_RPI.set_max_speed_hz(2000000)
        WS2801_RPI.flush()
        self.assertEqual(mock_spi.max_speed_hz, 2000000)

    def test_set_led_colors_buffer_list_multi_call(self, mock_spi):
        """Test set_led_colors_buffer_list_multi_call."""
        WS2801_RPI.set_number_of_leds()
        WS2801_RPI.set_led_colors_buffer_list_multi_call(1)
        WS2801_RPI.flush()
        WS2801_RPI.set_gamma(1)
        expected_result = ([255, 255, 255] + [0, 0, 0] *
                           (WS2801_RPI.get_number_of_leds() - 1))
        mock_spi.writebytes.assert_called_with(expected_result)
        with self.assertRaises(TypeError):
            WS2801_RPI.set_number_of_leds()
            WS2801_RPI.set_led_colors_buffer_list_multi_call("1")
        with self.assertRaises(TypeError):
            WS2801_RPI.set_number_of_leds()
            WS2801_RPI.set_led_colors_buffer_list_multi_call(1, 3)
        with self.assertRaises(TypeError):
            WS2801_RPI.set_number_of_leds()
            WS2801_RPI.set_led_colors_buffer_list_multi_call(["String"])
        with self.assertRaises(ValueError):
            WS2801_RPI.set_number_of_leds()
            WS2801_RPI.set_led_colors_buffer_list_multi_call([700])
        with self.assertRaises(ValueError):
            WS2801_RPI.set_number_of_leds()
            WS2801_RPI.set_led_colors_buffer_list_multi_call(700)
        with self.assertRaises(TypeError):
            WS2801_RPI.set_number_of_leds()
            WS2801_RPI.set_led_colors_buffer_list_multi_call([1, 3, 5],
                                                             [[255, 255, 255],
                                                             3])
        with self.assertRaises(TypeError):
            WS2801_RPI.set_number_of_leds()
            WS2801_RPI.set_led_colors_buffer_list_multi_call([1, 3, 5],
                                                             [[255, 255, 255],
                                                             [3]])
        with self.assertRaises(TypeError):
            WS2801_RPI.set_number_of_leds()
            WS2801_RPI.set_led_colors_buffer_list_multi_call([1, 3, 5],
                                                             [[255, 255, 255],
                                                             [3, 4, 6.7]])
        with self.assertRaises(ValueError):
            WS2801_RPI.set_number_of_leds()
            WS2801_RPI.set_led_colors_buffer_list_multi_call([1, 3, 5],
                                                             [[255, 255, 255],
                                                             [3, 4, 6700]])
        with self.assertRaises(TypeError):
            WS2801_RPI.set_number_of_leds()
            WS2801_RPI.set_led_colors_buffer_list_multi_call([1, 3, 5],
                                                             [255, 255]
                                                             )
        with self.assertRaises(TypeError):
            WS2801_RPI.set_number_of_leds()
            WS2801_RPI.set_led_colors_buffer_list_multi_call([1, 3, 5],
                                                             [255, 255, "String"]
                                                             )
        with self.assertRaises(ValueError):
            WS2801_RPI.set_number_of_leds()
            WS2801_RPI.set_led_colors_buffer_list_multi_call([1, 3, 5],
                                                             [255, 255, -3]
                                                             )

    @patch("WS2801_RPI.__logging")
    def test_set_get_led_colors_buffer_dict(self, mock_logger, mock_spi):
        WS2801_RPI.set_gamma(1)
        WS2801_RPI.set_number_of_leds(1)
        d = WS2801_RPI.get_led_colors_buffer_dict()
        self.assertEqual(d, {1: {"red": 0, "green": 0, "blue": 0}})
        d[1]["red"] = 27
        WS2801_RPI.set_led_colors_buffer_dict_multi_call(d)
        WS2801_RPI.flush()
        mock_spi.writebytes.assert_called_with([27, 0, 0])
        with self.assertRaises(TypeError):
            WS2801_RPI.set_led_colors_buffer_dict_multi_call([1, "red"])
        d[-1] = {"green": 255}
        with self.assertRaises(ValueError):
            WS2801_RPI.set_led_colors_buffer_dict_multi_call(d)
        del(d[-1])
        WS2801_RPI.set_number_of_leds(2)
        d[2] = "No dictionary"
        with self.assertRaises(TypeError):
            WS2801_RPI.set_led_colors_buffer_dict_multi_call(d)
        d[2] = {"red": 256}
        with self.assertRaises(ValueError):
            WS2801_RPI.set_led_colors_buffer_dict_multi_call(d)
        d[2] = {"green": 256}
        with self.assertRaises(ValueError):
            WS2801_RPI.set_led_colors_buffer_dict_multi_call(d)
        d[2] = {"blue": 256}
        with self.assertRaises(ValueError):
            WS2801_RPI.set_led_colors_buffer_dict_multi_call(d)
        del(d[2])
        d[4.0] = {"float": 3.14}
        WS2801_RPI.set_led_colors_buffer_dict_multi_call(d)
        self.assertTrue(mock_logger.warning.called)

    def test_set_spidev_bus_device(self, mock_spi):
        with self.assertRaises(TypeError):
            WS2801_RPI.set_spidev_bus_device("String", 0)
        WS2801_RPI.set_spidev_bus_device(0, 1)
        WS2801_RPI.flush()
        mock_spi.open.assert_called_with(bus=0, device=1)

    @patch("WS2801_RPI.time")
    def test_hello(self, mock_time, mock_spi):
        WS2801_RPI.hello()
        self.assertEqual(mock_spi.open.call_count, 6)
        self.assertEqual(mock_spi.writebytes.call_count, 6)

    @patch("WS2801_RPI.traceback")
    def test_flush(self, mock_traceback, mock_spi):
        mock_spi.open.side_effect = Exception("mock_spi raised an Excption")
        with self.assertRaises(Exception):
            WS2801_RPI.flush()

    def test_get_led_colors_buffer_list(self, mock_spi):
        WS2801_RPI.set_number_of_leds(5)
        WS2801_RPI.set_gamma(1)
        li = WS2801_RPI.get_led_colors_buffer_list()
        self.assertEqual(li, [[0, 0, 0]] * 5)


if __name__ == '__main__':
    unittest.main()


WS2801_RPI.py
from pkg_resources import get_distribution, DistributionNotFound
try:
    __version__ = get_distribution(__name__).version
except DistributionNotFound:
   # package is not installed
   pass

__doc__ = """
WS2801python_rpi: A python program to connect WS2801 driven LED strips.

Copyright (C) 2017  Markus Kupke <kupkemarkus@gmail.com>
Version: /version/

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

See <http://www.gnu.org/licenses/>.
""".replace("/version/", __version__)
import spidev as __spidev
import logging as __logging
import time
from timeit import default_timer as __timer
import traceback


__logging.warning(__doc__)

__BUS = 0
__DEVICE = 0
__MAX_SPEED_HZ = 976000
__NUMBER_LEDS = 0
__rgb_leds = []
__mode = 0b01
__last_flush = 0
# only for initialization. Is set to reasonable values at end of file
__gamma = [255]*256

# texts to separate into resources once I do a package (if ever)
__string_type_error_1 = "Attribute pixels must be an \
int addressing one of the available LEDs or a list of ints"
__string_type_error_2 = "Attribute rgb_values must be an \
list containing 3 ints between 0 and 255 or a list of lists with rgb values"
__string_type_error_3 = "Attribute must be a dictionary"
__string_type_error_4 = "Value must be a dictionary as well"
__string_type_error_6 = "bus and device must be either 0 or 1 each"
__string_value_error_1 = "You are addressing an LED that does not exist"
__string_value_error_2 = "rgb values must be between 0 and 255"
__string_runtime_error_2 = "Something weird happend: Buffer overflow"
__string_runtime_error_3 = "SPI Bus must be driven with clock speed > 1500HZ \
for WS2801"
__string_runtime_error_4 = "Mode must be between 0 and 3."
__string_warning_1 = "more leds addressed than rgb \
values given: assume last rgb for remaining leds"
__string_warning_3 = "Your dict contains floats as keys. \
These are not taken into account when adressing LEDs."
__string_warning_2 = "too many rgb values supplied: \
skipping"

# initialize SPI
# this should meet the requirements of WS2801
try:
    __spi = __spidev.SpiDev()
except:
    traceback.print_exc()
    raise


def set_spidev_bus_device(bus=0, device=0):
    """
    RASPBERRY PI Model 3 is normally connected via /dev/spidev0.0.

    If your model connects differently: adapt with this function.
    """
    if (not type(bus) is int or
            not type(device) is int or
            bus < 0 or
            bus > 1 or
            device < 0 or
            device > 1):
        raise TypeError(__string_type_error_6)
    global __BUS
    __BUS = bus
    global __DEVICE
    __DEVICE = device


def set_number_of_leds(leds=128):
    """
    Define the number of leds on your strip.

    This function has to be called if different number than 128!
    """
    global __NUMBER_LEDS
    __NUMBER_LEDS = leds
    global __rgb_leds
    __rgb_leds = [0] * __NUMBER_LEDS * 3


def get_number_of_leds():
    """Retrieve the number of LEDs set in the module."""
    return __NUMBER_LEDS


def hello():
    """Call one time to check LEDs."""
    for j in range(3):
        set_led_colors_buffer_list_multi_call(
                [i+1 for i in range(__NUMBER_LEDS)],
                [[255, 255, 255]] * __NUMBER_LEDS)
        flush()
        time.sleep(1)
        clear()
        flush()
        time.sleep(1)


def flush():
    """Send the bits to the leds. No parameters."""
    # WS2801 needs 500 micro seconds between flushes
    global __last_flush
    while (__timer() - __last_flush) <= 0.0005:
        pass
    try:
        __spi.open(bus=__BUS, device=__DEVICE)
        __spi.mode = __mode
        __spi.max_speed_hz = __MAX_SPEED_HZ
        __spi.writebytes(list(map(lambda rgb: __gamma[rgb], __rgb_leds)))
        __spi.close()
    except:
        traceback.print_exc()
        raise
    __last_flush = __timer()


def clear():
    """Switch all leds off. Needs a flush() to be active."""
    global __rgb_leds
    __rgb_leds[:] = [0 for i in range(__NUMBER_LEDS*3)]


def get_led_colors_buffer_dict():
    """
    Get the buffer for leds colors as dict.

    The leds are indexed with ints starting with 1 and values are dicts with
    rgb indexed with "red","green" and "blue"
    """
    global __NUMBER_LEDS
    leds_dict = {}
    rgb_dict = {"red": 0, "green": 0, "blue": 0}
    for index in range(0, __NUMBER_LEDS * 3, 3):
        rgb_dict = {"red": __rgb_leds[index],
                    "green": __rgb_leds[index+1],
                    "blue": __rgb_leds[index+2]}
        leds_dict[int((index+3)/3)] = rgb_dict
    return leds_dict


def set_led_colors_buffer_dict_multi_call(leds_dict):
    """
    Set the buffer for the leds via a dictionary.

    Key of dictionary is an int adressing the number of the led,
    all other keys are ignored.
    Value is a dictionary with the keys "red", "green" and "blue".
    Other values are ignored. If a color key is not found it defaults to 0.
    Values in the second dictinary can be ints between 0 and 255
    """
    if not type(leds_dict) is dict:
        raise TypeError(__string_type_error_3)
    for led, rgb_dict in leds_dict.items():
        rgb_list = [0, 0, 0]
        if type(led) is float:
            __logging.warning(__string_warning_3)
        if type(led) is int:
            if led < 1 or led > __NUMBER_LEDS:
                raise ValueError(__string_value_error_1)
            if not type(rgb_dict) is dict:
                raise TypeError(__string_type_error_4)
            for color, rgb_value in rgb_dict.items():
                if color == "red":
                    if (not type(rgb_value) is int or
                            rgb_value < 0 or
                            rgb_value > 255):
                        raise ValueError(__string_value_error_2)
                    rgb_list[0] = rgb_value
                if color == "green":
                    if (not type(rgb_value) is int or
                            rgb_value < 0 or
                            rgb_value > 255):
                        raise ValueError(__string_value_error_2)
                    rgb_list[1] = rgb_value
                if color == "blue":
                    if (not type(rgb_value) is int or
                            rgb_value < 0 or
                            rgb_value > 255):
                        raise ValueError(__string_value_error_2)
                    rgb_list[2] = rgb_value
            set_led_colors_buffer_list_multi_call(led, rgb_list)


def get_led_colors_buffer_list():
    """Get the list of rgb values for all LEDs."""
    return [[__rgb_leds[i], __rgb_leds[i+1],
             __rgb_leds[i+2]] for i in range(0, 3 * __NUMBER_LEDS, 3)]


def set_led_colors_buffer_list_multi_call(pixels, rgb_values=[255, 255, 255]):
    """Set a list of pixels to list of corresponding rgb_values (given as
    list of 3 values for red, green and blue).

    set_pixels([1,3,5],[[155,0,0],[0,0,0],[0,27,58]]) will set pixels
    1,3 and 5 to rgb values given.
    if length(rgb_values) < list(pixels) the last entry of rgb_values
    will count for the rest of the pixels.
    if pixel is a number only one pixel will be set.
    """
    # check input validity
    global __rgb_leds
    if not type(pixels) is int:
        if not type(pixels) is list:
            raise TypeError(__string_type_error_1)
        for val in pixels:
            if (not type(val) is int):
                raise TypeError(__string_type_error_1)
            if val < 1 or val > __NUMBER_LEDS:
                raise ValueError(__string_value_error_1)
    else:
        if pixels > __NUMBER_LEDS or pixels < 1:
            raise ValueError(__string_value_error_1)
    if not type(rgb_values) is list:
        raise TypeError(__string_type_error_2)
    if type(rgb_values[0]) is list:
        for rgb_value in rgb_values:
            if not type(rgb_value) is list:
                raise TypeError(__string_type_error_2)
            if not len(rgb_value) == 3:
                raise TypeError(__string_type_error_2)
            for i in rgb_value:
                if (not type(i) is int):
                    raise TypeError(__string_type_error_2)
                if i < 0 or i > 255:
                    raise ValueError(__string_value_error_2)
    else:
        if not len(rgb_values) == 3:
            raise TypeError(__string_type_error_2)
        for i in rgb_values:
            if (not type(i) is int):
                raise TypeError(__string_type_error_2)
            if i < 0 or i > 255:
                raise ValueError(__string_value_error_2)

    # write into __rgb_leds
    if type(pixels) is int:
        pixels = [pixels]
    if type(rgb_values[0]) is int:
        tmp = []
        tmp.append(rgb_values)
        rgb_values = tmp
    for i, val in enumerate(pixels):
        if i < len(rgb_values):
            rgb = rgb_values[i]
        __rgb_leds[(val-1)*3:(val)*3] = rgb

    # some final sanity checks
    if len(__rgb_leds) != __NUMBER_LEDS*3:
        raise RuntimeError(__string_runtime_error_2)
    if len(pixels) > len(rgb_values):
        __logging.warning(__string_warning_1)
    if len(pixels) < len(rgb_values):
        __logging.warning(__string_warning_2)


def set_max_speed_hz(hz):
    """
    Set the communication speed for SPI.

    Default: 976000HZ. There is no need to call this function if default is ok.
    Consult devspi docu https://github.com/doceme/py-spidev.
    """
    if hz <= 1500:
        # 500 micro sec clock low is the time WS2801 needs to reset. hence
        # it must be driven with a clock higher than 1000hz otherwise it will
        # reset all the time :-)
        raise ValueError(__string_runtime_error_3)
    global __MAX_SPEED_HZ
    __MAX_SPEED_HZ = hz


def set_gamma(gamma=2.1):
    """Set the gamma correction."""
    for i in range(256):
        __gamma[i] = int(((i/255.0)**gamma) * 255.0)


# initialize the module with 128 leds
set_gamma()
set_number_of_leds()
